File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque

n,m = map(int,input().split())

graph = [list(map(int,input())) for i in range(n)]

def bfs(gy,gx):
    myqueue = deque()
    myqueue.append((gx,gy))
    dx = [ -1, 1, 0, 0]
    dy = [ 0, 0, 1, -1]
    while myqueue:
        x,y = myqueue.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0<=nx<m and 0<=ny<n:
                if graph[ny][nx] == 0:
                    continue
                if graph[ny][nx] == 1:
                    graph[ny][nx] = graph[y][x]+1
                    myqueue.append((nx,ny))

    return graph[n-1][m-1]
print(bfs(0,0))

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        // BufferedReader 선언
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
//        StringTokenizer st ; // = new StringTokenizer(br.readLine());
        String[] split = br.readLine().split(" ");
        int gy = Integer.parseInt(split[0]);
        int gx = Integer.parseInt(split[1]);
        int[][] graph = new int[gy][gx];

        for (int i =0; i <gy; i++) {
            String st = br.readLine();
            for (int j =0; j < gx; j++) {
                graph[i][j] = st.charAt(j) - '0';
            }
        }
        Deque<Integer> result = new ArrayDeque<>();
        Deque<int[]> myque = new ArrayDeque<>();
//        for (int i = 0; i < gy; i++) {
//            for (int j = 0; j < gx; j++) {
//                if (graph[i][j] == 1) {
//                    int bfs = bfs(graph, i, j, gy, gx, myque);
//                    result.add(bfs);
//                }
//            }
//        }
        bfs(graph, 0,0,gy,gx,myque);
        System.out.println(graph[gy-1][gx-1]);

    }
    public static int bfs(int[][] graph,int y, int x,int gy, int gx,Deque<int[]> myque ) {
        myque.add(new int[]{y,x});
        graph[y][x] = 1;
        int count = 1;
        while (!myque.isEmpty()) {
            int[] cur = myque.poll();
            y = cur[0];
            x = cur[1];
            int [] dy = {0,0,1,-1};
            int [] dx = {1,-1,0,0};
            for (int i = 0; i < 4; i++) {
                int ny = y + dy[i];
                int nx = x + dx[i];
                if (0<=ny && ny<gy && 0<= nx && nx<gx ) {
                    if (graph[ny][nx] == 1) {
                        graph[ny][nx] = graph[y][x] +1;
                        myque.add(new int[]{ny,nx});
                        count +=1;
                    }

                }
            }
        }
        return count;
    }
}